### ISSUE TITLE
👍Unified Language Model Pre-training for Natural Language Understanding and Generation

### ISSUE THREAD

**Issue by fury00812** [fury00812/Survey#4]
Unified Language Model Pre-training for Natural Language Understanding and Generation  
Li Dong, Nan Yang, Wenhui Wang, Furu Wei, Xiaodong Liu, Yu Wang, Jianfeng Gao, Ming Zhou, Hsiao-Wuen Hon  
ACL2019  
https://arxiv.org/abs/1905.03197

## 一言で
MicrosoftがBERTを基に作った事前学習モデル. マスク行列を導入することで言語理解と言語生成, 両タスクへの適用を可能にした. 

# Abstract
- 言語理解と言語生成の両タスクに適用可能な, 統一的(Unified)事前学習モデルの提案
- 事前学習には3種類の言語モデル(単方向LM/双方向LM/seq2seqLM)を使用
- 理解 : GLUEベンチマークや, SQuAD2.0やCoQAの質問応答でSOTA
- 生成 : 質問文の生成でSOTA. 更に生成した文でData AugumentationすることによりSQuADのスコアがアップ.

# 1	Introduction
- 様々なタスクに適応できる, 事前学習済み言語モデルが流行っている(BERTとか)
- 既存の事前学習モデルは, タスクに応じてベースの言語モデルが異なる
  - ELMo : 2つの単方向LM(based on LSTM)
  - GPT : 単方向Transformer
  - BERT : 双方向Transformer

![0](https://user-images.githubusercontent.com/35480446/59548328-ab133e00-8f88-11e9-98bc-c772bc6577d1.png)

- BERTは自然言語理解タスクに対し非常に有用だが, 自然言語生成タスクへの適用は困難
- 本稿では多様な下流タスクに適用可能な事前学習モデルUNILMを提案
- <font color="Green">事前学習時, 3つの言語モデルはマスク単語の予測を行う. この時モデル毎に異なる方法でマスキングする. パラメータは共有する. </font>
  - 単方向LM : left-to-rightならマスクされた右側の単語を, 左側の全ての単語を使って予測
  - 双方向LM : 普通のMasked LM(BERTの)
  - seq2seqLM : attention Encoder-Decoder のような働き. 生成タスクを可能にする

- メリット
1. 単体で複数のタスクにfine-tuneできるので複数モデルを用意する必要が無くて楽
1. 異なる言語モデルでパラメータを共有して事前学習するのでよりgeneralな言語表現が得られる
1. seq2seq LMの導入により言語生成が可能に

# 2	Unified Language Model Pre-training
- inputは, Position/Segment/Token Embeddingの和（これはBERTと同様）
- inputを~~3つの言語モデル（多層Transformer, パラメータは共通）に与え, それぞれ事前学習~~
マスク行列M（後述）を変える事で3つの言語モデルを学習するのと等価になる

![1](https://user-images.githubusercontent.com/35480446/59548333-b7979680-8f88-11e9-8ca8-5dc4b8d3bd8c.png)

## 2.1	Input Representation
- 文はサブワードを用いてトークナイズする

## 2.2	Backbone Network: Transformer
- モデルにはdeep transformerを利用
- Mはマスク行列 : 双方向LMの場合は0行列

## 2.3	Pre-training Objectives
- トークンをランダムに選択して[MASK]に置き換える
- 4つのタスクによって[MASK]を予測

### 2.3.1	Unidirectional LM
- left-to-right : [MASK]の左側の単語列のみから予測
e.g. $"x_1 x_2 [MASK] x_4"$から$x_1,x_2$を使って$[MASK]$を予測
- right-to-left : [MASK]の右側の単語列のみから予測

### 2.3.2	Bidirectional LM
- [MASK]の予測に前後全ての単語を使用
- マスク行列Mは0行列

### 2.3.3	Sequence-to-Sequence LM
- Segment1には前後から参照可能
- Segment2にはSegment2に属する単語からしか参照出来ずしかも先の情報は見れない
- <font color="Red">**seq2seqでencoderがdecoderのattentionを見れないのと同じ**</font>
- e.g. 入力文$t_1t_2$とその次の文$t_3t_4t_5$があるとする. この時モデルに$"[SOS] t_1 t_2 [EOS] t_3 t_4 t_5 [EOS]"$を入力
$t_1$や$t_2$を予測する時は$[SOS], t_1, t_2, [EOS]$を使う

### 2.3.4 Next Sentence Prediction
- BERT同様, 隣接文予測も事前学習する. これはBidirectional LMにやらす

## 2.4 Pre-training Setup
- 学習時には、バッチを3等分して各モデルに割り当てる. 
- 24層Transformer, 隠れ層1024次元, パラメータ総数はおよそ340M!
	- これはDevlinらの$BERT_{large}$(2018)に準拠
- 事前学習データ : Wikipedia,BookCorpus(English)
- 語彙数28996, 入力シーケンスの最大長は512
- マスキング確率15%
	- 選ばれたトークンは, 80:10:20の確率で[MASK]:ランダムトークン:そのまま
	- トークンを選ぶ時, 20%の確率でbigramかtrigram単位でマスクする

## 2.5 Fine-tuning on Downstream Tasks
- 自然言語理解ならBidirectional LMを, 言語生成ならSequence-to-Sequence LMを強化（マスク行列Mで操作できる）

# 3	Experiments
1. 文書要約タスク(CNN/Daily Mail)でSOTA
![2](https://user-images.githubusercontent.com/35480446/59548361-355ba200-8f89-11e9-99f3-47fc2d711218.png)

2. 質問応答タスク(SQuAD,CoQA)でSOTA
- SQuAD
![3](https://user-images.githubusercontent.com/35480446/59548363-41dffa80-8f89-11e9-96cb-128ded9a916d.png)
- CoQA
![4](https://user-images.githubusercontent.com/35480446/59548368-4ad0cc00-8f89-11e9-85e1-51bd657a025f.png)

3. 質問文生成タスク(data:SQuAD)でSOTA
![5](https://user-images.githubusercontent.com/35480446/59548371-53c19d80-8f89-11e9-93b0-ed67bdb8f558.png)

4. 3で生成したデータでSQuADを水増し->スコアアップ
![6](https://user-images.githubusercontent.com/35480446/59548375-5de39c00-8f89-11e9-9b3a-acfae7c7228a.png)

5. GLUEでもSOTA
![7](https://user-images.githubusercontent.com/35480446/59548381-6fc53f00-8f89-11e9-8310-54a85f3116b6.png)
